File: frontend/src/types/chart.ts
import { z } from "zod";

// Chart type enum
export const ChartTypeSchema = z.enum([
  "line",
  "area",
  "bar",
  "stackedArea",
  "scatter",
  "pie",
]);
export type ChartType = z.infer<typeof ChartTypeSchema>;

// Value formatter type
export const FormatterTypeSchema = z.enum([
  "usd",
  "percent",
  "date",
  "datetime",
  "number",
]);
export type FormatterType = z.infer<typeof FormatterTypeSchema>;

// Series definition
export const SeriesSchema = z.object({
  key: z.string(),
  label: z.string(),
  color: z.string().nullish(), // nullish = null | undefined
});
export type Series = z.infer<typeof SeriesSchema>;

// Axis configuration
export const AxisConfigSchema = z.object({
  label: z.string().nullish(),
  formatter: FormatterTypeSchema.nullish(),
});
export type AxisConfig = z.infer<typeof AxisConfigSchema>;

// Main ChartSpec
export const ChartSpecSchema = z.object({
  version: z.literal(1),
  type: ChartTypeSchema,
  title: z.string().nullish(),
  data: z.array(z.record(z.string(), z.unknown())),
  xKey: z.string(),
  series: z.array(SeriesSchema),
  xAxis: AxisConfigSchema.nullish(),
  yAxis: AxisConfigSchema.nullish(),
  showLegend: z.boolean().nullish(),
  showTooltip: z.boolean().nullish(),
  height: z.number().nullish(),
});
export type ChartSpec = z.infer<typeof ChartSpecSchema>;

File: src/models/chart.py
"""Chart specification models for frontend visualization."""

from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel


class ChartType(str, Enum):
    """Supported chart types."""

    LINE = "line"
    AREA = "area"
    BAR = "bar"
    STACKED_AREA = "stackedArea"
    SCATTER = "scatter"
    PIE = "pie"


class FormatterType(str, Enum):
    """Value formatter types for axis display."""

    USD = "usd"
    PERCENT = "percent"
    DATE = "date"
    DATETIME = "datetime"
    NUMBER = "number"


class Series(BaseModel):
    """A data series definition."""

    key: str
    label: str
    color: Optional[str] = None


class AxisConfig(BaseModel):
    """Axis configuration."""

    label: Optional[str] = None
    formatter: Optional[FormatterType] = None


class ChartSpec(BaseModel):
    """
    Chart specification that the frontend can render.

    This schema matches the frontend's Zod validation schema.
    """

    version: int = 1
    type: ChartType
    title: Optional[str] = None
    data: list[dict[str, Any]]
    xKey: str
    series: list[Series]
    xAxis: Optional[AxisConfig] = None
    yAxis: Optional[AxisConfig] = None
    showLegend: Optional[bool] = True
    showTooltip: Optional[bool] = True
    height: Optional[int] = 300
